lib/predicate/validate.js

```diff
@@ -0,0 +1,68 @@
+function getType(v) {
+    return Object.prototype.toString.call(v)
+        .replace(/\[object\s(\w+)\]/, '$1').toLowerCase();
+}
+
+function validateAllFields(spec, candidate, prefix, errors, ctx) {
+    const specType = getType(spec);
+    const candidateType = getType(candidate);
+
+    if (specType === 'regexp') {
+        if (!spec.test(candidate)) {
+            errors.push(`${prefix}: invalid string pattern`);
+        }
+        return;
+    }
+
+    if (specType === 'function') {
+        if (!spec.call(ctx, candidate)) {
+            errors.push(`${prefix}: failed predicate`);
+        }
+        return;
+    }
+
+    if (candidateType !== specType) {
+        errors.push(`${prefix}: want ${specType}, have ${candidateType}`);
+        return;
+    }
+
+    switch (specType) {
+    case 'array':
+        candidate.forEach((v, i) => {
+            validateAllFields(spec[0], v, `${prefix}.${i}`, errors, ctx);
+        });
+        return;
+    case 'object':
+    case 'error':
```

Comment:
`Object.prototype.toString.call(new Error())` (the input to the switch statement)  returns `[object Error]`, which is then lowercased. Fallthrough is intentional.